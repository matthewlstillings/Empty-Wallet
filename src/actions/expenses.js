import uuid from 'uuid';
import database from '../firebase/firebase.js';

//Add Expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            notes = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = {description, notes, amount, createdAt};
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
               id: ref.key,
               ...expense 
            }));
        });
    };
};


//Edit Expense 
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//Start Edit Expense *Start* means for firebase data
export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then((ref) => {
            dispatch(editExpense(id, updates));
        });
    };
};

//Remove Expense 
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//Start Remove Expense

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch, getState) => {
       const uid = getState().auth.uid;
       return database.ref(`users/${uid}/expenses/${id}`).remove().then((ref) => {
           dispatch(removeExpense({ id }));
       });
    };
};

// Set Expenses
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
}); 

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {  //.once() grabs the value of data just one time
            const expenses = [];
            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });    
    };
};