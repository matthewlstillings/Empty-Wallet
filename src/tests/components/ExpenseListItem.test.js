import React from 'react';
import {shallow} from 'enzyme';
import Expense from '../../components/ExpenseListItem';
import expenses from '../fixtures.js/expenses';

test('should render expense item', () => {
    const wrapper = shallow(<Expense {...expenses[1]}/>);
    expect(wrapper).toMatchSnapshot();
});

