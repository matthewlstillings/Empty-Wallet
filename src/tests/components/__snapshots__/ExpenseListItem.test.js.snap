// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense item 1`] = `
<li>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      apple
    </h3>
  </Link>
  <p>
    $1,095.00
  </p>
  <p>
    December 27th, 1969
  </p>
</li>
`;
